Packages/com.unity.inputsystem/InputSystem/API/InputBindingsCollection.cs

```diff
@@ -0,0 +1,203 @@
+#if UNITY_INPUT_SYSTEM_ENABLE_GLOBAL_ACTIONS_API
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using UnityEngine.InputSystem.Utilities;
+
+namespace UnityEngine.InputSystem
+{
+    /// <summary>
+    /// A collection that can intelligently iterate over the bindings of an input action.
+    /// </summary>
+    /// <remarks>
+    /// Input bindings are stored such that normal bindings, composite bindings, and composite part bindings
+    /// e.g. the 'negative' and 'positive' bindings of a 1D axis composite, or the 'up', 'down', 'left', 'right'
+    /// bindings of a 2D vector composite, are all stored together in a single contiguous array. This collection
+    /// enables easy iteration over, or direct indexing into, the bindings of a specified Input Action without
+    /// having to manually keep track of whether a binding is a composite part or a normal binding.
+    ///
+    /// The iterator and indexer of this type return BindingWithIndex instances
+    ///
+    /// Note that an Input Actions' bindings can change due to binding re-resolution, which can be caused by
+    /// something as simple as adding a new binding, or a device connecting or disconnecting, so it is not
+    /// advised to store instances of this type as the underlying data can change.
+    /// </remarks>
+    public readonly struct InputBindingsCollection : IEnumerable<BindingWithIndex>
+    {
+        private readonly InputAction m_InputAction;
+        private readonly ReadOnlyArray<InputBinding> m_Bindings;
+        private readonly EnumerationBehaviour m_EnumerationBehaviour;
+
+        public InputBindingsCollection(InputAction inputAction, EnumerationBehaviour enumerationBehaviour)
+        {
+            Debug.Assert(inputAction != null);
+
+            m_InputAction = inputAction;
+            m_Bindings = m_InputAction.bindings;
+            m_EnumerationBehaviour = enumerationBehaviour;
+        }
+
+        /// <summary>
+        /// An indexer into the bindings collection that respects the enumeration behaviour.
+        /// </summary>
+        /// <param name="index"></param>
+        /// <returns></returns>
+        /// <exception cref="ArgumentOutOfRangeException"></exception>
+        /// <exception cref="InvalidOperationException"></exception>
+        public BindingWithIndex this[int index]
+        {
+            get
+            {
+                if (index < 0 || index > m_Bindings.Count)
+                    throw new ArgumentOutOfRangeException(nameof(index));
+
+                if (m_EnumerationBehaviour == EnumerationBehaviour.SkipCompositeParts)
+                {
+                    var currentIndex = 0;
+                    var i = 0;
+                    while (i < m_Bindings.Count)
+                    {
+                        if (currentIndex == index)
+                            return new BindingWithIndex(m_Bindings[currentIndex],
```

Comment:
This line is tricky to parse, perhaps it can be made simpler to read, at minimum currentIndex could be replaced with index for less text as they are equal here.